### ISSUE TITLE
Is it possibe to use different device name on different node

### ISSUE THREAD

**Issue by garyyang85** [ceph/ceph-ansible#1671]
Assume I have two node for osd deployment.
On one osd, I have device **sda sdb**; on the second node , I have device **sdc sdd**. Is it possible to use ceph-ansible deploy osd on these two nodes at one time?  In group_vars/osds.yml, it seems that we assume both osd node have the same device name?

**Comment by gfidente** [comment]
+1

**Comment by fultonj** [comment]
@guits @leseb not having the feature described here introduces a regression for TripleO integration in its current state. I am curious on your thoughts on it. 

In TripleO a user can compose a different role for a different type of OSD service, [e.g. server1 with sda/sdb and server2 with sdc/sdd/sde (the different counts are relevant for this example)](https://github.com/RHsyseng/hci/blob/master/other-scenarios/mixed-nodes/ceph-mixed.yaml#L15), and Heat will pass different parameter lists to puppet-ceph because Heat is orchestrating; not Ansible. 

If ceph-ansible supports a wild card syntax for specifying OSDs, e.g. /dev/sd*, that might be a way, provided you can also exclude patterns. That said, there's the rule of thumb that unbalanced OSD daemons across different nodes won't perform as well and should be avoided.

**Comment by laggeduout** [comment]
Would it not be viable to use osd_auto_discovery in this case?

**Comment by guits** [comment]
what about declaring devices in inventory host file for earch osd host ?

**Comment by fultonj** [comment]
@guits and @laggeduout thanks for your suggestions. I talked to @gfidente about it too and he saw a connection with node-specific hieradata and using it to populate the device inventory file that we could look into to address this on the TripleO side. I also like the osd_auto_discovery option. 

TL;DR
TripleO supports modifying the default node configuration by role [1] and node-specific hieradata [2]. They can both be used to achieve the same goal. It sounds like we could make a different implementation of the node specific hiera data on the TripleO side. Not by using hiera, but by allowing the user to specify in heat a similar syntax to denote a disk [2] for a different Ironic node. We could then pass that information to ceph-ansible and do a translation in Mistral so that it appears in our dynamic inventory file. The tricky part is doing extra queries to Ironic to line up the IP the host got in the inventory, but I don't think it's too bad. Something to research. 

[1] http://tripleo.org/install/advanced_deployment/node_config.html
[2] http://tripleo.org/install/advanced_deployment/node_specific_hieradata.html

**Comment by andrewschoen** [comment]
@fultonj I would probably use ``host_vars`` for this, but defining devices in your inventory would work as well.  Here's some upstream ansible docs that might help explain better: http://docs.ansible.com/ansible/latest/intro_inventory.html#splitting-out-host-and-group-specific-data

I don't believe ``osd_auto_discovery`` is very well tested, I'd advise against using it.

**Comment by leseb** [comment]
As discussed, declaring particular devices for each host in your inventory file is the way to go, closing this.

**Comment by garyyang85** [comment]
@leseb @andrewschoen Thanks for the solution.  Could you give an example configuration for the case in issue description?

**Comment by guits** [comment]
@garyyang85 , something like this should work :

```
[mons]
mon0_host
mon1_host
mon2_host

[osds]
osd0_host devices="['/dev/sda', '/dev/sdb', '/dev/sdc']"
osd1_host devices="['/dev/sdx', '/dev/sdy', '/dev/sdz']"
osd2_host devices="['/dev/sda', '/dev/sdb', '/dev/sdc']"
```

**Comment by garyyang85** [comment]
Thank you very much!!

**Comment by hasonhai** [comment]
hello @guits , how's about the lvm case?
I have a group of nodes with vg named centos, and another group with vg named rhel. Is there a way to specify them in the inventory file?

**Comment by hasonhai** [comment]
Thanks! It works!
I left the sample here, later someone may need.
```
[osds]
DEFR4APP[501:506] lvm_volumes="[{'data': 'ceph-osd', 'data_vg': 'centos', 'journal': 'ceph-journal', 'journal_vg': 'centos'}]"
DEFR4APP[510:512] lvm_volumes="[{'data': 'ceph-osd', 'data_vg': 'rhel', 'journal': 'ceph-journal', 'journal_vg': 'rhel'}]"
```

**Comment by caisan** [comment]
@hasonhai @guits  hi, guys, have you tried to add an extra osd disk on OSD node via ceph-ansible playbook ? does the infrastructure-playbooks/add-osd.yml works  for you ? what is the procedure ?
(infrastructure-playbooks/add-osd.yml not works for me, maybe i do the wrong procedure )

**Comment by hasonhai** [comment]
@caisan have you copied the file add-osd.yml to the parent directory? some files in infrastructure-playbooks directory can not play directly.
